### DISCUSSION TITLE
OSError: could not find or load spatialindex_c-64.dll

### DISCUSSION THREAD

**Discussion by msaifp** [pyinstaller/pyinstaller#6060] [Help]
File "CSV_to_Shapefile_Converter.py", line 11, in
File "", line 1007, in _find_and_load
File "", line 986, in _find_and_load_unlocked
File "", line 680, in load_unlocked
File "PyInstaller\loader\pyimod03_importers.py", line 546, in exec_module
File "geopandas_init.py", line 1, in
File "", line 1007, in _find_and_load
File "", line 986, in _find_and_load_unlocked
File "", line 680, in _load_unlocked
File "PyInstaller\loader\pyimod03_importers.py", line 546, in exec_module
File "geopandas_config.py", line 126, in
File "geopandas_config.py", line 112, in _default_use_pygeos
File "", line 1007, in _find_and_load
File "", line 986, in _find_and_load_unlocked
File "", line 680, in _load_unlocked
File "PyInstaller\loader\pyimod03_importers.py", line 546, in exec_module
File "geopandas_compat.py", line 202, in
File "", line 1007, in _find_and_load
File "", line 986, in _find_and_load_unlocked
File "", line 680, in load_unlocked
File "PyInstaller\loader\pyimod03_importers.py", line 546, in exec_module
File "rtree_init.py", line 9, in
File "", line 1007, in _find_and_load
File "", line 986, in _find_and_load_unlocked
File "", line 680, in _load_unlocked
File "PyInstaller\loader\pyimod03_importers.py", line 546, in exec_module
File "rtree\index.py", line 6, in
File "", line 1007, in _find_and_load
File "", line 986, in _find_and_load_unlocked
File "", line 680, in _load_unlocked
File "PyInstaller\loader\pyimod03_importers.py", line 546, in exec_module
File "rtree\core.py", line 75, in
File "rtree\finder.py", line 67, in load
OSError: could not find or load spatialindex_c-64.dll
[2904] Failed to execute script 'CSV_to_Shapefile_Converter' due to unhandled exception!

Hi, this error has been driving me nuts for weeks now. I have been trying to package my script which works fine in a Spyder Python IDE but when I try to use Pyinstaller to package it into an executable file, it throws the aforementioned error. I can't seem to resolve this issue as I've tried all solutions on the web - particularly those detailed in the Geopandas forums (#1812). Any help to resolve this issue would be greatly appreciated!

**Comment by rokm** [comment]
Try adding `--collect-binaries rtree` to your PyInstaller command.

**Reply by msaifp** [reply]
Hi Rokm,

I have tried adding this command to the PyInstaller command but I still get the same output result of OSError: could not find or load spatialindex_c-64.dll.

**Reply by rokm** [reply]
After you added the `--collect-binaries rtree`, does you `dist\<name>` contain `rtree\lib` directory that contains `spatialindex_c-64.dll` and `spatialindex-64.dll`?

**Reply by ougx** [reply]
Having the same problem: `rtree` is not included in `dist`

```python
pyinstaller --hidden-import rtree --collect-binaries rtree wellPlot.py
```

**Reply by rokm** [reply]
`--collect-binaries rtree` should do the trick with python.org python and `rtree` wheel from PyPI...

If you forgot to mention that you're using anaconda, then it's a bit more complicated, because conda installs those DLLs into its environment's `Libraries\bin` and patches `rtree.finder` to look in there.

The finder still seems to look into `rtreee\lib` as a fallback (as well as at location pointed to by `SPATIALINDEX_C_LIBRARY` environment variable).

So you can find those two DLLs in your conda environment and ensure they are collected into `rtree\lib` in the `dist` folder (either via spec file, or copying them manually if you're doing `onedir` build).

Or you *could* create a custom hook for rtree that uses:
```python
from PyInstaller.utils.hooks.conda import collect_dynamic_libs
binaries = collect_dynamic_libs('rtree')
```
but this will collect the DLLs into top-level application directory, so you would also need to ensure in your program that `SPATIALINDEX_C_LIBRARY` points there before attempting to import `rtree` (or importing whatever ends up importing `rtree`), i.e.:
```python
import sys

if getattr(sys, 'frozen', False):
    import os
    os.environ['SPATIALINDEX_C_LIBRARY'] = sys._MEIPASS
    
import rtree  # or geopandas or whatever
```

**Reply by ougx** [reply]
Thanks for the detailed solution. I do use `conda`. I copied the dll to `rtreee\lib` but it did not work.

`collect_dynamic_libs('rtree')` produced empty list.

**Reply by rokm** [reply]
> Thanks for the detailed solution. I do use `conda`. I copied the dll to `rtreee\lib` but it did not work.

Did the error change or remain the same (i.e., `OSError: could not find or load spatialindex_c-64.dll`)?

> `collect_dynamic_libs('rtree')` produced empty list.

Just to clarify, did you use `PyInstaller.utils.hooks.conda.collect_dynamic_libs` or `PyInstaller.utils.hooks.collect_dynamic_libs`? For `conda`, the former should return list of DLLs that includes `spatialindex_c-64.dll`, whereas the latter should return empty list (unless you are using PyPI wheel in a conda environment).

**Reply by ougx** [reply]
> > Thanks for the detailed solution. I do use `conda`. I copied the dll to `rtreee\lib` but it did not work.
> 
> Did the error change or remain the same (i.e., `OSError: could not find or load spatialindex_c-64.dll`)?

My bad. It is actually another error:
```python
AttributeError: partially initialized module 'fiona' has no attribute '_loading' (most likely due to a circular import)
```
> 
> > `collect_dynamic_libs('rtree')` produced empty list.
> 
> Just to clarify, did you use `PyInstaller.utils.hooks.conda.collect_dynamic_libs` or `PyInstaller.utils.hooks.collect_dynamic_libs`? For `conda`, the former should return list of DLLs that includes `spatialindex_c-64.dll`, whereas the latter should return empty list (unless you are using PyPI wheel in a conda environment).

Yes, I used the wrong `collect_dynamic_libs`. `PyInstaller.utils.hooks.conda.collect_dynamic_libs` did return a list of dll tuples. 

So for the `fiona` error I found this solution of changing the `fiona` code: https://stackoverflow.com/questions/67865273/pyinstaller-fiona-has-no-attribute-loading-most-likely-due-to-a-circular
Is there other way to fix it?
Thanks!

**Reply by rokm** [reply]
Hmm, a `--collect-submodules fiona` seemed to do the trick here (for both 1.8.13 that was installed by default, and for 1.8.20 from conda-forge).

**Reply by ougx** [reply]
Thanks, it solved the problem! So to sum up, below is what I do to use `geopandas` in `conda` env in Windows:

```cmd
pyinstaller --collect-submodules fiona --collect-data geopandas -y wellPlot.py

mkdir dist\wellPlot\rtree\lib
copy /y c:\Miniconda3\Library\bin\spatialindex*.dll dist\wellPlot\rtree\lib\
```

Thanks for your help @rokm

**Reply by msaifp** [reply]
Thank you for the detailed solution @rokm . This has now solved my isse with rtree and the spatialindex_c-64.dll not being found.

**Reply by rokm** [reply]
For the record, the `rtree` DLLs should be automatically picked up in the future, thanks to pyinstaller/pyinstaller-hooks-contrib#291.
